protocols/relay/src/v2/client.rs

```diff
@@ -146,11 +152,40 @@ impl Client {
 }
 
 impl NetworkBehaviour for Client {
-    type ConnectionHandler = handler::Prototype;
+    type ConnectionHandler = Either<Handler, dummy::ConnectionHandler>;
     type OutEvent = Event;
 
-    fn new_handler(&mut self) -> Self::ConnectionHandler {
-        handler::Prototype::new(self.local_peer_id, None)
+    fn new_handler(
+        &mut self,
+        peer: &PeerId,
+        connected_point: &ConnectedPoint,
+    ) -> Result<Self::ConnectionHandler, Box<dyn std::error::Error + Send + 'static>> {
+        if connected_point.is_relayed() {
+            if let Some(event) = self.initial_events.remove(peer) {
+                log::debug!(
+                    "Established relayed instead of direct connection to {:?}, \
+                     dropping initial in event {:?}.",
+                    peer,
+                    event
```

Comment:
How do we know that there is no second **direct** connection in progress of establishment to the given peer? How do we know that this event was predetermined for this connection and not another connection still being established?

This is rather complex. I rewrote this comment 4 times. I might be missing something.

Reply:
We can index the `HashMap` by the dialed address instead of the peer ID. That should fix things I guess?